Auto-switch to a mapping/agglomerate if precomputed mesh uses that

daniel-wer opened the pull request:
I wrapped the two menu items which can be used to load a precomputed mesh with an HOC that asks the user to activate the corresponding mapping for the currently active mesh file before loading the mesh. The mapping activation is then triggered automatically and awaited, before the mesh loading is triggered.
The backend was adapted to send the designated mapping name for each mesh file if available or `null` otherwise.

Additionally, I performed some refactoring by moving mapping related code to a saga and extending the `isMappingEnabled` status to an enum with 3 states: DISABLED, ACTIVATING, ENABLED allowing easier selection of the layer relevant for json mappings and allowing to find out when a mapping activation has actually finished (with data reloading for agglomerate mappings).

### TODO:
- [x] ~~Confirm disappears in segment list if the segment list is rerendered which happens while hovering (not sure how to elegantly solve this)~~ No longer happens for me, not sure why :shrug: 
- [x] Test with a mesh file that has no mapping_name

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Open a dataset with a mesh file that was computed for a specific mapping and contains that information in the hdf5 metadata. For example this one:
[meshfile.zip](https://github.com/scalableminds/webknossos/files/7603183/meshfile.zip)
- Right click in a viewport and select "Load Mesh (precomputed)". You should be asked to confirm that the corresponding mapping will be activated. If confirmed, the mapping should be activated and the correct mesh should be loaded. If denied, nothing should happen.
  - Do the same from the segment list.
- Merger mode should still work.
- Manual (json) mapping activation should still work.

### Issues:
- fixes #5563
- contributes to https://github.com/scalableminds/webknossos/issues/5695#:~:text=mapping%20setup%20is,work%20right%20now.

------
(Please delete unneded items, merge only when none are left open)
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Needs datastore update after deployment
- [x] Ready for review

daniel-wer commented on frontend/javascripts/oxalis/model/bucket_data_handling/mappings.js:
```diff
@@ -46,34 +45,8 @@ class Mappings {
-    app.vent.listenTo(app.vent, "webknossos:ready", this.registerReloadHandler);
   }
 
-  registerReloadHandler = () => {
```
I moved this code to the mapping_saga, see `maybeReloadData`

philippotto reviewed (commented):
Beautiful stuff! I'll test once all todo's are done :)

daniel-wer commented on frontend/javascripts/oxalis/model/sagas/mapping_saga.js:
```diff
@@ -24,18 +25,55 @@ import {
+};
+
 export default function* watchActivatedMappings(): Saga<void> {
+  const oldActiveMappingByLayer = {};
```
Agreed, but since we're using action channels I'd argue that the initialization should stay in line 45 since immediately after that the action channels start picking up actions.

philippotto replied:
Ok, sounds good :+1:

daniel-wer commented on frontend/javascripts/oxalis/model/sagas/mapping_saga.js:
```diff
@@ -24,18 +25,55 @@ import {
+  const mapping = getMappingInfo(activeMappingByLayer, layerName);
+
+  const isAgglomerateMappingInvolved = isAgglomerate(oldMapping) || isAgglomerate(mapping);
+  const hasChanged = oldMapping !== mapping;
```
Good catch!

daniel-wer commented on frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view_helper.js:
```diff
@@ -124,3 +129,67 @@ export async function loadMeshFromFile(
+        visible={isConfirmVisible}
+        onConfirm={async () => {
+          setConfirmVisible(false);
+          Store.dispatch(setMappingAction(layerName, currentMeshFile.mappingName, "HDF5"));
+          await waitForCondition(
+            () =>
+              getMappingInfo(
+                Store.getState().temporaryConfiguration.activeMappingByLayer,
+                layerName,
+              ).mappingStatus === MappingStatusEnum.ENABLED,
+            100,
+          );
```
I'll leave it as-is for now, since resolving the deferred would not be straight-forward. Essentially, the mapping activation is a bit convoluted and consists of multiple actions that are partly triggered by a saga and the deferred would need to be dragged along to be available to be resolved in the very end.

philippotto replied:
Ok, makes sense :+1:

daniel-wer commented:
@philippotto Thank you for your feedback :) I incorporated that and took care of the remaining Todos, so the PR should be ready for testing.

There's one more thing I thought of:
- [ ] Ideally, we would distinguish between mesh files without any mapping information (because they were created before the mapping_name was added to the file) and mesh files with a mapping_name that is empty (because the mesh file was computed for the over segmentation). An empty mapping_name is signified by an hdf5 `Empty` attribute (see https://docs.h5py.org/en/stable/high/dataset.html?highlight=empty#creating-and-reading-empty-or-null-datasets-and-attributes). The server should send `undefined` for mesh files without a mapping_name and `null` for mesh files with an empty mapping_name. This way, the confirm would never appear if the mapping_name is `undefined` regardless of whether a mapping is activated or not.

philippotto reviewed (commented):
Great stuff! Testing went well after I solved some local (unrelated) hiccups :+1: 

I only noticed two minor UI issues:
- I added the mesh file to a dataset which didn't have any mappings (mea culpa). The prompt said "Mapping with name agglomerate_view_70 and type HDF5 does not exist. Available mappings are ." <-- "There are no available mappings." would be nicer.
- The popconfirm gets really "wide" on my screen. Can we give it a max-width: 500 or something like this?

>  Ideally, we would distinguish between mesh files without any mapping information (because they were created before the mapping_name was added to the file) and mesh files with a mapping_name that is empty (because the mesh file was computed for the over segmentation). An empty mapping_name is signified by an hdf5 Empty attribute (see https://docs.h5py.org/en/stable/high/dataset.html?highlight=empty#creating-and-reading-empty-or-null-datasets-and-attributes). The server should send undefined for mesh files without a mapping_name and null for mesh files with an empty mapping_name. This way, the confirm would never appear if the mapping_name is undefined regardless of whether a mapping is activated or not.

I agree that this would be necessary probably. So, this is blocking the PR, right?

daniel-wer commented:
Thank you for your feedback. I addressed the two issues you mentioned and also fixed a race condition when activating a mapping with an invalid name or type (for example, via an edited sharing link).

> I agree that this would be necessary probably. So, this is blocking the PR, right?

Yes, maybe @fm3 can have a look at that once he is back in office?

fm3 commented:
>  The server should send undefined for mesh files without a mapping_name and null for mesh files with an empty mapping_name.

Sure, I can have a look and see if jhdf5 can distinguish between the two variants. Could you supply me with a meshfile for each, so I can test this? Also, JSON does not have a concept of undefined, so I assume I should omit the field in this case, and write null for the empty name?

philippotto commented:
> [...] so I assume I should omit the field in this case, and write null for the empty name?

This should work :+1: Another idea would be to also supply a `version` field in the JSON payload. Not sure if we already have some versioning for the mesh file? The back-end could append the version field by inspecting the `mapping_name` field. Not sure, whether that's more elegant..

daniel-wer commented:
I prepared two mesh files, one without the `mapping_name` attribute and one with an empty one:
[meshfile_no_mapping_name.zip](https://github.com/scalableminds/webknossos/files/7703186/meshfile_no_mapping_name.zip)
[meshfile_empty_mapping_name.zip](https://github.com/scalableminds/webknossos/files/7703188/meshfile_empty_mapping_name.zip)

Either of the two methods should work (mesh files without the mapping_name have no artifact_schema_version attribute and mesh files with a mapping_name have the artifact_schema_version attribute set to 1). I slightly prefer the omit/null version since it's less complex, but no strong opinion.

fm3 commented:
I just tested what the existing code does, and indeed it implements the omit/null version. If that is fine with you, let’s leave it as is :)  

Return value for me with the files above was 
```
[
    {
        "meshFileName": "meshfile_no_mapping_name"
    },
    {
        "meshFileName": "meshfile_empty_mapping_name",
        "mappingName": null
    }
]
```

I’ll just add a comment in the back-end about this case (we don’t usually expect null anywhere).

daniel-wer commented:
> I just tested what the existing code does, and indeed it implements the omit/null version. If that is fine with you, let’s leave it as is :)

I have adapted the frontend accordingly and have confirmed that this works as expected.

@philippotto Would be great if you could have a final look :)

philippotto reviewed (approved):
Terrific! Code looks :+1: Didn't test again, since I think your testing should be enough :)
